numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
Here's my concern about "option 2": my understanding is that we eventually settled on supporting two kinds of objects: duck-arrays where binops and ufuncs do the same thing (objects with a real `__numpy_ufunc__`, and special objects that define their own binops and for which ufuncs are undefined (objects where `__numpy_ufunc__ = None`). Because "Option 2" classes return `NotImplemented` from binops even when the other object doesn't set `__numpy_ufunc__ = None`, they make it possible to define classes to have contradictory binop and ufunc definitions. Sometimes, sorta, depending on which array types they interact with. Plus it's just harder to reason about or manage `__numpy_ufunc__` when a `NotImplemented` return value is overloaded to poke two different complicated dispatch systems in different ways?

Regarding the error message: doing `a + b` and getting some sort of error message like `unsupported operands for <ufunc 'add'>` seems pretty reasonable to me? I think most people will work out the relationship between `+` and `add` and it gives the hint that `+` is dispatching to the ufunc, which seems like a useful thing to let people know.

Regarding backports: This doesn't seem like a good reason to me to change the semantics of code in numpy? If projects want to take advantage of `__numpy_ufunc__`, then can declare a dependency on `numpy >= 1.13` and it's not really a big deal these days with conda and wheels and all that; or they can borrow your "option 2" code even without it being included with numpy. I'd rather that the version we actually ship match what ndarray does.

Regarding `__numpy_ufunc__` implementations that ignore `__numpy_ufunc__ = None`: here you're distinguishing between two variants of option 2, and I'm in favor of option 1 (which always has to check for `__numpy_ufunc__ = None` in binops), so I won't respond to the details. But... the reminder that `__numpy_ufunc__` implementations might choose to accept classes that have `__numpy_ufunc__ = None` is an important point, I think! It can also lead to weird situations where a class can have `__numpy_ufunc__ = None` and override binops, yet also sometimes participate in ufuncs and give different results. What I'm wondering is if we should enforce the `__numpy_ufunc__ = None` rule by changing the ufunc dispatch to be:
- first, check if any dispatch-relevant arguments have `__numpy_ufunc__ = None`. If so, raise `TypeError` immediately.
- otherwise, [the current dispatch rules, subclasses before superclasses left to right]

?